### PR TITLE
chore(common): xml2js -> fast-xml-parser

### PR DESCRIPTION
**Pull request by srl295**
replace xml2js with [fast-xml-parser](https://www.npmjs.com/package/fast-xml-parser) (much better support)

Fixes: #12208

@keymanapp-test-bot skip

### CODE REVIEW

File: common/schemas/kvks/kvks.schema.json
Comment (mcdurdin) [inline common/schemas/kvks/kvks.schema.json:1]: Note that schema changes need to be kept in sync with corresponding files in api.keyman.com (we really should automate this...)

File: developer/src/common/web/utils/src/types/kpj/kpj-file-reader.ts [hunk 1]
```diff
@@ -13,20 +13,19 @@ export class KPJFileReader {
...
+    attributeNamePrefix: '', // to avoid '@_' prefixes
+    numberParseOptions: {
+      skipLike: /(?:)/, // parse numbers as strings
+      hex: null,
+      leadingZeros: null,
+      eNotation: null,
+    },
+  });
+    const raw = parser.parse(file.toString());
+    delete raw['?xml']; // XML prologue
```
Comment (mcdurdin) [inline developer/src/common/web/utils/src/types/kpj/kpj-file-reader.ts:27]: I wonder if this is going to be the same across all our parsers, and if we should have a wrapper?

File: developer/src/common/web/utils/package-lock.json
Comment (mcdurdin) [inline developer/src/common/web/utils/package-lock.json:1]: We shouldn't have a package-lock.json here as this is part of a workspace.

**Comment by mcdurdin** [comment]
@SabineSIL FYI

File: common/schemas/kvks/kvks.schema.json
Comment (srl295) [reply common/schemas/kvks/kvks.schema.json:1]: Hm, where would I find those files? This particular schema change is not a change to the actual XML, just of the in-memory JSON format.

File: developer/src/common/web/utils/src/types/kpj/kpj-file-reader.ts (hunk 1 above)
Comment (srl295) [reply developer/src/common/web/utils/src/types/kpj/kpj-file-reader.ts:27]: maybe, though there are some nuances such as for the ldml test data. So I'm not sure it needs to be a wrapper.  Though a utility function to knock out the prologue might make sense.

File: common/schemas/kvks/kvks.schema.json
Comment (mcdurdin) [reply common/schemas/kvks/kvks.schema.json:1]: https://github.com/keymanapp/api.keyman.com/tree/master/schemas

File: developer/src/common/web/utils/src/types/kpj/kpj-file-reader.ts (hunk 1 above)
Comment (mcdurdin) [reply developer/src/common/web/utils/src/types/kpj/kpj-file-reader.ts:27]: yup, we can let the dust settle first

File: common/schemas/kvks/kvks.schema.json
Comment (srl295) [reply common/schemas/kvks/kvks.schema.json:1]: OK, i see the `$` there.  So after this PR merges, copy it over.  Now the kvks schema is versioned as `10.0` - does this schema need a bump? It's not actually a change to the kvks file itself.

Comment (mcdurdin) [reply common/schemas/kvks/kvks.schema.json:1]: I _think_ we can keep it the same version but we should update the documentation to clarify that it is used only with fast-xml-parser and not xml2js.

**Comment by srl295** [comment]
I'd like to refactor the common parsing, etc as a followon PR after we make sure this is actually working due to the dependency change.

**Review by mcdurdin — CHANGES_REQUESTED** [review]
I've started the review but have a significant change request which will, I hope, reduce the impact and scope. As it stands, I think the changes would have a number of negative cascading impacts on consumers of the data, and so I would prefer to have the data coming out of the parser match the current shape, so we don't have to change anything downstream from the parser.

File: developer/src/common/web/utils/src/types/kpj/kpj-file-reader.ts [hunk 2]
```diff
@@ -13,20 +13,20 @@ export class KPJFileReader {
...
-    parser.parseString(file, (e: unknown, r: unknown) => {
-      if(e) {
-        throw e;
-      }
-      data = r as KPJFile;
-    });
+  const parser = new XMLParser({
+    htmlEntities: true,
+    ignoreAttributes: false, // We'd like attributes, please
+    attributeNamePrefix: '', // to avoid '@_' prefixes
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply developer/src/common/web/utils/src/types/kpj/kpj-file-reader.ts:19]: See later comment for bigger request that makes this suggestion null

**Comment by srl295** [comment]
> I've started the review but have a significant change request which will, I hope, reduce the impact and scope. As it stands, I think the changes would have a number of negative cascading impacts on consumers of the data, and so I would prefer to have the data coming out of the parser match the current shape, so we don't have to change anything downstream from the parser.

Some seem to be positive changes, but make some sense, basically `xml2js: true` mode.

**Comment by mcdurdin** [comment]
> Some seem to be positive changes

Oh for sure, but negative in that they require lots of changes to consumers of the API, which is costly.

File: developer/src/common/web/utils/src/types/kps/kps-file.ts [hunk 3]
```diff
@@ -91,22 +91,20 @@ export interface KpsFileLanguages {
...
-    /**
-     * relationship between this package and the related package, "related" is default if not specified
-     */
-    Relationship?: "deprecates";
-  }
+  ID: string;
+  /**
+   * relationship between this package and the related package, "related" is default if not specified
+   */
+  Relationship?: "deprecates";
```
Comment (ermshiperete) [inline developer/src/common/web/utils/src/types/kps/kps-file.ts:107]: Shouldn't this be

```suggestion
  Relationship?: "deprecates"|"related";
```

if the default is "related"? (I didn't check if there are other possible values)

Comment (mcdurdin) [reply developer/src/common/web/utils/src/types/kps/kps-file.ts:107]: Currently only `"deprecates"` is allowed:

https://github.com/keymanapp/keyman/blob/400444933cb51e972d7c9b95fbaa06b520d8feb6/common/schemas/kps/kps.xsd#L108-L109

**Comment by mcdurdin** [comment]
did you mean to close this?

**Comment by srl295** [comment]
> did you mean to close this?

Yes, in favor of one based on #12482 (I guess I could reopen this once sorted)

**Comment by mcdurdin** [comment]
> > did you mean to close this?
> 
> Yes, in favor of one based on #12482 (I guess I could reopen this once sorted)

OK, np. Please make sure you document when you do actions like this because I was quite confused!

Shall I'll go ahead and delete the branch?

**Comment by srl295** [comment]
> Shall I'll go ahead and delete the branch?

Not yet because I am using it to work from! 

I will either reopen it or open a new one

**Pull request closed without merging**